### PR TITLE
[WIP] `inserted` might be None, fixes #1525

### LINKED ISSUE #1525: `inserted` might be None
See https://github.com/Kinto/kinto/blob/a6b08389ebb441bee10f336fa03627eb3a8aa37f/kinto/core/storage/postgresql/__init__.py#L306-L308

If `inserted` becomes `None` you get a `TypeError` trying to call `inserted['inserted']`.

This has actually happened in a real production instance:

```
Exception(most recent call first)

TypeError: 'NoneType' object is not subscriptable
  File "kinto/core/utils.py", line 414, in follow_subrequest
    return request.invoke_subrequest(subrequest, **kwargs), subrequest
  File "pyramid/router.py", line 195, in invoke_subrequest
    return self.invoke_request(request, _use_tweens=use_tweens)
  File "pyramid/router.py", line 249, in invoke_request
    response = handle_request(request)
  File "pyramid/router.py", line 156, in handle_request
    view_name
  File "pyramid/view.py", line 642, in _call_view
    response = view_callable(context, request)
  File "pyramid/config/views.py", line 181, in __call__
    return view(context, request)
  File "pyramid/viewderivers.py", line 390, in attr_view
    return view(context, request)
  File "pyramid/viewderivers.py", line 368, in predicate_wrapper
    return view(context, request)
  File "pyramid/viewderivers.py", line 301, in secured_view
    return view(context, request)
  File "pyramid/viewderivers.py", line 439, in rendered_view
    result = view(context, request)
  File "pyramid/viewderivers.py", line 148, in _requestonly_view
    response = view(request)
  File "cornice/service.py", line 494, in wrapper
    response = view_()
  File "kinto/core/resource/__init__.py", line 468, in put
    record = self.model.create_record(new_record)
  File "kinto/core/resource/model.py", line 315, in create_record
    record = super().create_record(record, parent_id)
  File "kinto/core/resource/model.py", line 191, in create_record
    auth=self.auth)
  File "statsd/client.py", line 37, in _wrapped
    return_value = f(*args, **kwargs)
  File "kinto/core/storage/postgresql/__init__.py", line 308, in create
    if not inserted['inserted']:
```

For those with access see [this Sentry entry](https://sentry.prod.mozaws.net/operations/buildhub-prod/issues/3163658/).

### PR DESCRIPTION
**Pull request by peterbe**
Fixes #1525

- [ ] Add documentation.
- [ ] Add tests.
- [ ] Add a changelog entry.
- [ ] Add your name in the contributors file.
- [ ] If you changed the HTTP API, update the [API_VERSION](https://github.com/Kinto/kinto/blob/master/kinto/__init__.py#L15) constant and add an API changelog entry [in the docs](https://github.com/Kinto/kinto/blob/master/docs/api/index.rst)
- [ ] If you added a new configuration setting, update the [`kinto.tpl`](https://github.com/Kinto/kinto/blob/master/kinto/config/kinto.tpl) file with it.

### CODE REVIEW

**Comment by peterbe** [comment]
@leplatrem I need your help to finish this PR. Lots of help :)

I've only just started getting into kinto dev. Not sure where to add a test or even what should happen. In https://github.com/Kinto/kinto/issues/1525#issuecomment-381739832 I outlined a very rough way to trigger the `insert is None` case but I did it with pure psycopg2 and not in the context of kinto. Also, I suspect the unit test, to trigger the race condition, needs to use a threading which I don't know if/how we already do that. 

In your hands for guidance.

**Comment by peterbe** [comment]
(it's my intention to return to the checkboxes in the PR description once I get going)

File: kinto/core/storage/postgresql/__init__.py [hunk 1]
```diff
@@ -305,6 +305,29 @@ def create(self, collection_id, parent_id, record, id_generator=None,
...
+            #  create(my_collection_id, my_parent_id, {
+            #      'id': 'myid',
+            #      'last_modified': '2018-04-16 06:53:12.005'
+            #  })
+            # (note the only difference is the different last_modified)
+            # ...then the above SQL query will return 0 rows and thus 'inserted'
+            # becomes None.
+            #
+            # Basically, the client called .create() that second time wrongly.
+            raise exceptions.UnicityError(id_field, record)
```
[the remark this answers is not in the document]
Comment (peterbe) [reply kinto/core/storage/postgresql/__init__.py]: How would I get that record? Fetching it? E.g. `SELECT data FROM records where id=:object_id AND parent_id=:parent_id AND collection_id=:collection_id` ?

[the remark this answers is not in the document]
Comment (Natim) [reply kinto/core/storage/postgresql/__init__.py]: You can probably use self.get(collection_id, parent_id, object_id) but that's the idea I guess yes.

**Comment by leplatrem** [comment]
> Not sure where to add a test or even what should happen.

We have 2 places for storage tests:
- `kinto/core/storage/testing.py` for tests that apply to every implementations of storage backends
- `tests/core/test_storage:PostgreSQLStorageTest` for postgres specific tests

So a test along this one would work:
https://github.com/Kinto/kinto/blob/9e22784dc57291c8f93624f8c1175cabd9b7f080/tests/core/test_storage.py#L259-L271
You can setup all the threading / raw pyscopg you need, let me know if you can't figure out how to transpose your snippet to the kinto ecosystem

Also, I realize that when the storage backend raises a `UnicityError` we don't automatically transform it into a `409` like we do for `IntegrityError` here:
https://github.com/Kinto/kinto/blob/9e22784dc57291c8f93624f8c1175cabd9b7f080/kinto/core/views/errors.py#L112
Maybe we should also change that so that unicity errors are returned as 409. Either in the global error view as above for any uncaught exception, or add a `try/except` in the resource code:
https://github.com/Kinto/kinto/blob/9e22784dc57291c8f93624f8c1175cabd9b7f080/kinto/core/resource/__init__.py#L345
https://github.com/Kinto/kinto/blob/9e22784dc57291c8f93624f8c1175cabd9b7f080/kinto/core/resource/__init__.py#L468

> Note: This is slightly related to #1407

File: kinto/core/storage/postgresql/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (glasserc) [reply kinto/core/storage/postgresql/__init__.py]: Doing a `self.get()` leads to a different race condition. What happens if another thread deletes the record in between your insert failing and your `get()`?

**Comment by peterbe** [comment]
Before I dig deeper into this, I'm excited about this idea (of mine): https://github.com/Kinto/kinto/issues/1407#issuecomment-382489691
Basically, instead of doing fancy upsert CTE functions, we probably ought to use proper row-level locking and rely on the primary key integrity checking.

**Comment by leplatrem** [comment]
What's the status of this?

File: kinto/core/storage/postgresql/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (peterbe) [reply kinto/core/storage/postgresql/__init__.py]: @glasserc Is it worth taking the risk perhaps?
If the first thread (who manages to make the successful insert) does a delete right afterwards, they better do so quickly. Really quickly :)

What I'm alluding to is that swallowing that risk and doing something like:
```python
was_there = self.get(collection_id, parent_id, object_id)
raise UnicityErrror(was_there)
```
...will be a pretty decent solution.

**Comment by leplatrem** [comment]
Closing in favor #1919

**Pull request closed without merging**